table/partitioned_filter_block.cc

```diff
@@ -9,76 +9,62 @@
-      index_block_builder_(index_block_restart_interval),
-      index_per_partition_(index_per_partition),
-      index_block_restart_interval_(index_block_restart_interval) {
-  sub_index_builder_ =
-      CreateIndexBuilder(sub_type_, comparator_, prefix_extractor_,
-                         index_block_restart_interval_, index_per_partition_);
-}
 
-PartitionIndexBuilder::~PartitionIndexBuilder() { delete sub_index_builder_; }
+PartitionedFilterBlockBuilder::PartitionedFilterBlockBuilder(
+    const SliceTransform* prefix_extractor, bool whole_key_filtering,
+    FilterBitsBuilder* filter_bits_builder, int index_block_restart_interval,
+    PartitionedIndexBuilder* const p_index_builder)
+    : FullFilterBlockBuilder(prefix_extractor, whole_key_filtering,
+                             filter_bits_builder),
+      index_on_filter_block_builder_(index_block_restart_interval),
+      p_index_builder_(p_index_builder) {}
+
+PartitionedFilterBlockBuilder::~PartitionedFilterBlockBuilder() {}
 
-void PartitionIndexBuilder::AddIndexEntry(
-    std::string* last_key_in_current_block,
-    const Slice* first_key_in_next_block, const BlockHandle& block_handle) {
-  sub_index_builder_->AddIndexEntry(last_key_in_current_block,
-                                    first_key_in_next_block, block_handle);
-  num_indexes++;
-  if (UNLIKELY(first_key_in_next_block == nullptr)) {  // no more keys
-    entries_.push_back({std::string(*last_key_in_current_block),
-                        std::unique_ptr<IndexBuilder>(sub_index_builder_)});
-    sub_index_builder_ = nullptr;
-  } else if (num_indexes % index_per_partition_ == 0) {
-    entries_.push_back({std::string(*last_key_in_current_block),
-                        std::unique_ptr<IndexBuilder>(sub_index_builder_)});
-    sub_index_builder_ =
-        CreateIndexBuilder(sub_type_, comparator_, prefix_extractor_,
-                           index_block_restart_interval_, index_per_partition_);
+// FullFilterBlockBuilder
+void PartitionedFilterBlockBuilder::MaybeCutAFilterBlock() {
+  if (!p_index_builder_->ShouldCutFilterBlock()) {
+    return;
   }
+  filter_gc.push_back(std::unique_ptr<const char[]>(nullptr));
+  Slice filter = filter_bits_builder_->Finish(&filter_gc.back());
+  std::string& index_key = p_index_builder_->GetPartitionKey();
+  filters.push_back({index_key, filter});
+}
+
+void PartitionedFilterBlockBuilder::AddKey(const Slice& key) {
+  MaybeCutAFilterBlock();
+  filter_bits_builder_->AddKey(key);
 }
 
-Status PartitionIndexBuilder::Finish(
-    IndexBlocks* index_blocks, const BlockHandle& last_partition_block_handle) {
-  assert(!entries_.empty());
-  // It must be set to null after last key is added
-  assert(sub_index_builder_ == nullptr);
-  if (finishing == true) {
-    Entry& last_entry = entries_.front();
+Slice PartitionedFilterBlockBuilder::Finish(
+    const BlockHandle& last_partition_block_handle, Status* status) {
+  if (finishing_filters == true) {
```

Comment:
It took me some time to verify the logic of this method. Would this be a little bit more clear?
```
if (filter.empty() && !finishing_filters) {
  return Slice();
} else if (!filter.empty()) {
  auto filter = filters.front();
  filters.pop_front();
  finish_filters = true;
  index_builder.Add(filter.key);
  return filter.filter;
} else {
  return index_builder.Finish();
}